### PR TITLE
JAMES-1950 Log duration of imap & jmap requests

### PR DESCRIPTION
**Pull request by aduprat**
Commit cea76d4 is not part of this review (#654).

### CODE REVIEW

File: metrics/metrics-logger/src/main/java/org/apache/james/metrics/logger/DefaultTimeMetric.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Override
+    public String name() {
+        return name;
+    }
+
+    @Override
+    public long elapseTimeInMs() {
+        return stopwatch.elapsed(TimeUnit.MILLISECONDS);
```
Comment (quynhn) [inline metrics/metrics-logger/src/main/java/org/apache/james/metrics/logger/DefaultTimeMetric.java]: Sorry but I would like to ask again. Do we need to call stop method before calculation?

Comment (chibenwa) [inline metrics/metrics-logger/src/main/java/org/apache/james/metrics/logger/DefaultTimeMetric.java]: Nope

**Comment by aduprat** [comment]
I wanted this PR to be simple in order to be merged, i will do another PR on top of it with your recommendations.

**Comment by aduprat** [comment]
test this please

File: metrics/metrics-logger/src/main/java/org/apache/james/metrics/logger/DefaultTimeLogger.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.metrics.logger;
+
+import org.apache.james.metrics.api.TimeLogger;
+import org.apache.james.metrics.api.TimeMetric;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultTimeLogger implements TimeLogger {
```
Comment (mbaechler) [inline metrics/metrics-logger/src/main/java/org/apache/james/metrics/logger/DefaultTimeLogger.java]: why do we mix metrics and logging ?

File: mpt/impl/imap-mailbox/core/pom.xml [hunk 3]
```diff
@@ -66,6 +66,10 @@
             <artifactId>james-server-mailbox-adapter</artifactId>
         </dependency>
         <dependency>
+            <groupId>org.apache.james</groupId>
+            <artifactId>metrics-logger</artifactId>
```
Comment (mbaechler) [inline mpt/impl/imap-mailbox/core/pom.xml]: should be metrics-api or I am missing something ?

File: metrics/metrics-logger/src/main/java/org/apache/james/metrics/logger/DefaultTimeLogger.java (hunk 2 above)
Comment (aduprat) [inline metrics/metrics-logger/src/main/java/org/apache/james/metrics/logger/DefaultTimeLogger.java]: The name is may be not well chosen, the idea is more to have a time publisher.

File: mpt/impl/imap-mailbox/core/pom.xml (hunk 3 above)
Comment (aduprat) [inline mpt/impl/imap-mailbox/core/pom.xml]: No, metrics-logger is one implementation of metrics-api.
In MPT, i've bounded the simple, and only one yet, implementation.

Comment (mbaechler) [inline mpt/impl/imap-mailbox/core/pom.xml]: ah, ok, that's for mpt, I missed that

File: metrics/metrics-logger/src/main/java/org/apache/james/metrics/logger/DefaultTimeLogger.java (hunk 2 above)
Comment (mbaechler) [inline metrics/metrics-logger/src/main/java/org/apache/james/metrics/logger/DefaultTimeLogger.java]: IMO it should be handle as a metric so we can put it in a dashboard and consumed by a logger if we want to.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesMethod.java [hunk 4]
```diff
@@ -57,18 +64,22 @@
 
     public Stream<JmapResponse> process(JmapRequest request, ClientId clientId, MailboxSession mailboxSession) {
         Preconditions.checkArgument(request instanceof SetMessagesRequest);
+        TimeMetric timeMetric = timeMetricFactory.generate(METHOD_NAME.getName());
         try {
-            return Stream.of(
+            Stream<JmapResponse> responses = Stream.of(
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesMethod.java]: Can't you still return stream.of() ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetVacationResponseMethod.java [hunk 5]
```diff
@@ -72,20 +79,25 @@ public SetVacationResponseMethod(VacationRepository vacationRepository, Notifica
...
-            return Stream.of(JmapResponse
+            Stream<JmapResponse> responses = Stream.of(JmapResponse
                 .builder()
                 .clientId(clientId)
                 .error(ErrorResponse.builder()
                     .type(INVALID_ARGUMENTS1)
                     .description(INVALID_ARGUMENT_DESCRIPTION)
                     .build())
                 .build());
+            timeLogger.log(timeMetric);
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetVacationResponseMethod.java]: Why no try / finally here ?

Could we get consistant ?

What if I also want to know response times on invalid requests ?

**Comment by aduprat** [comment]
see #665

**Pull request closed without merging**